finality-aleph/src/validator_network/io.rs

```diff
@@ -0,0 +1,176 @@
+use std::fmt::{Display, Error as FmtError, Formatter};
+
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+
+use crate::validator_network::Data;
+
+// We allow sending up to 16MiB, that should be enough forever.
+pub const MAX_DATA_SIZE: u32 = 16 * 1024 * 1024;
+
+/// A general error when sending or receving data.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    ConnectionClosed,
+    DataTooLong(u32),
+}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use Error::*;
+        match self {
+            ConnectionClosed => write!(f, "connection unexpectedly closed"),
+            DataTooLong(length) => write!(
+                f,
+                "encoded data too long - {} bytes, the limit is {}",
+                length, MAX_DATA_SIZE
+            ),
+        }
+    }
+}
+
+/// An error when sending data.
+#[derive(Debug, PartialEq, Eq)]
+pub struct SendError(Error);
+
+impl Display for SendError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<Error> for SendError {
+    fn from(e: Error) -> Self {
+        SendError(e)
+    }
+}
+
+/// An error when receiving data.
+#[derive(Debug, PartialEq, Eq)]
+pub enum ReceiveError {
+    Error(Error),
+    DataCorrupted,
+}
+
+impl Display for ReceiveError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use ReceiveError::*;
+        match self {
+            Error(e) => write!(f, "{}", e),
+            DataCorrupted => write!(f, "received corrupted data"),
+        }
+    }
+}
+
+impl From<Error> for ReceiveError {
+    fn from(e: Error) -> Self {
+        ReceiveError::Error(e)
+    }
+}
+
+/// Sends some data using the stream.
+pub async fn send_data<S: AsyncWriteExt + Unpin, D: Data>(
+    mut stream: S,
+    data: D,
+) -> Result<S, SendError> {
+    let mut encoded = data.encode();
+    let len = u32::try_from(encoded.len()).map_err(|_| Error::DataTooLong(u32::MAX))?;
+    if len > MAX_DATA_SIZE {
+        return Err(Error::DataTooLong(len).into());
+    }
+    let mut payload = len.to_le_bytes().to_vec();
+    payload.append(&mut encoded);
```

Comment:
We explicitly wanted this to work on `Data`, because keeping typing information everywhere is convenient.

How would we avoid allocations if this was declared for a type? Types often don't have constant encoding length.

Reply:
Making low level network code depend on so many (unnecessary from network's perspective) traits from `Data` doesn't look very convenient. If that would be a method for some type, then it could have some internal buffer (potentially even shared between connections). The  `Codec` trait was invented partially around idea of limiting number of necessary allocations (writing to `Output` buffer instead of transforming into Vec<u8>). In this approach it could be simply switched with `From<[u8]>` and `Into<[u8]>'.